runtime/ai/file_search.go

```diff
@@ -0,0 +1,151 @@
+package ai
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/modelcontextprotocol/go-sdk/mcp"
+	"github.com/rilldata/rill/runtime"
+)
+
+const SearchFilesName = "search_files"
+
+type SearchFiles struct {
+	Runtime *runtime.Runtime
+}
+
+var _ Tool[*SearchFilesArgs, *SearchFilesResult] = (*SearchFiles)(nil)
+
+type SearchFilesArgs struct {
+	Pattern     string `json:"pattern" jsonschema:"The pattern to search for. Supports regular expressions."`
+	CaseSensitive bool `json:"case_sensitive,omitempty" jsonschema:"Whether the search should be case-sensitive. Defaults to false."`
+	GlobPattern string `json:"glob_pattern,omitempty" jsonschema:"Optional glob pattern to filter files (e.g., '**.yaml', '**models/**'). If not provided, searches all files."`
+}
+
+type SearchFilesResult struct {
+	Matches []SearchMatch `json:"matches"`
+}
+
+type SearchMatch struct {
+	Path    string   `json:"path"`
+	Lines   []int    `json:"lines"`
+	Snippets []string `json:"snippets"`
+}
+
+func (t *SearchFiles) Spec() *mcp.Tool {
+	return &mcp.Tool{
+		Name:        SearchFilesName,
+		Title:       "Search files",
+		Description: "Searches for a pattern across files in the Rill project. Returns matching file paths, line numbers, and snippets. Use this before read_file to discover which files contain specific content.",
+		Meta: map[string]any{
+			"openai/toolInvocation/invoking": "Searching files...",
+			"openai/toolInvocation/invoked":  "Searched files",
+		},
+	}
+}
+
+func (t *SearchFiles) CheckAccess(ctx context.Context) (bool, error) {
+	return checkDeveloperAgentAccess(ctx, t.Runtime)
+}
+
+func (t *SearchFiles) Handler(ctx context.Context, args *SearchFilesArgs) (*SearchFilesResult, error) {
```

Comment:
The implementation looks good, would be nice to have a test for it as well. See `runtime/ai/metrics_view_query_test.go` for how to write a tool-specific test: https://github.com/rilldata/rill/blob/e03c48f9466e0a35f2b026a9dd25f8fe5f251a44/runtime/ai/metrics_view_query_test.go#L11-L50